core/cap-0046-07.md

```diff
@@ -40,272 +36,405 @@ The fee structure is designed to discourage "spam" traffic and overall waste of
 
-Changes to ledger data:
-```
+The total resource consumption for every one the 'competitive' resources must not exceed the ledger-wide limits. The specific limits are specified in sections below on per-resource basis.
 
-TBD: `LedgerEntry` with expiration date
+The usual `GeneralizedTransactionSet` validity and comparison rules also apply to Soroban corresponding to the semantics described in [CAP-0042](./cap-0042.md).
 
-```
+#### Transaction validation
 
-### Semantics
+All Soroban transactions must have `ext.sorobanData()` extension present and populated.
 
-Validity constraints:
-* source account must be able to pay for the total fee bid `tx.fee` for that transaction.
-* the number of smart contract transactions cannot exceed `ledgerMaxTxCount` per ledger
+`resources` contain the declared values of resources that the transaction is paying the fee for. These values have to not exceed the limits specified by the network settings. 
 
-#### Resources with contention and inclusion fees
+`resourceFee` is computed based on the `resources` declared in `tx` and transaction envelope size:
 
-The approach taken in this proposal is to decompose fees into:
-* `resourcesFee` - derived from how much resources a transaction is using and network parameters that evolve with time as to price those resources appropriately
-* `inclusionFeeBid` - this is the "social value" part of the fee, it represents the intrinsic value that the submitter puts on that transaction.
+`resourceFee(tx) = Instructions_fee(resources.instructions) + LedgerDataAccess_fee(resources) + NetworkData_fee(size(txEnvelope)) + Historical_flat_fee(size(txEnvelope))`
 
-The resources where we allow competition are:
-* instructions (virtual instructions to execute)
-* ledger data access (bytes transferred)
-* network propagation (bandwidth)
+Note, that `Historical_flat_fee` is  a 'competitive' resource, but its constant for any transaction execution result and thus is a part of non-refundable fee (as its refund is always 0).
 
-Transactions will have to pay both for `resourcesFee` and `inclusionFeeBid`, with `inclusionFeeBid` used to prioritize transactions relative to each other (both when flooding and for inclusion in transaction sets).
+`refundableFee` corresponds to the `refundableResourcesFee` component and has to be not lower than the initial refundable fee derived from `resources`:
 
-#### TransactionSet semantics
+`sorobanData.refundableFee >= initialRefundableFee(tx)`
 
-While transactions bid specific `inclusionFeeBid`, the effective bid may be lowered within a transaction set component by setting `inclusionFee`.
+where
 
-When set:
-* all transactions within the component must bid more than `inclusionFee`, ie for each transaction `inclusionFeeBid >= inclusionFee`
-* the effective inclusion bid for transactions in that group is `inclusionFee`
+`initialRefundableFee(tx) = Events_fee(resources.contractEventsSizeBytes)`
```

Comment:
Should this mention something about the return value?

Reply:
Updated.